[TSK-723] [TSK-724] [TSK-725] ui-kit: GeneralCard, GeneralCardCarousel

ObitoDevv opened the pull request:
## Description

<!-- Provide a clear and concise description of the changes made in this pull request for UI-Kit components. -->

- Component name: `GeneralCard` , `GeneralCardCarousel`
- Figma link or screenshot: [Figma Design](https://www.figma.com/design/8KEwRuOoD5IgxTtFAtLlyS/Just_Do_Ad-1.2?node-id=6340-115938&t=IrCukPhd3jcco8aU-4)

![image](https://github.com/user-attachments/assets/800164a2-d329-42e7-9dfc-2cffbbcbb041)


![Image](https://github.com/user-attachments/assets/abd72c2c-100f-4d8a-9457-cc8f68d914f3)

---

## Checklist

- [x] I have run the lint, build, and test commands locally, as explained in the README, before creating this PR to ensure the code is clean and functional.
- [x] My changes align with the **core models** that have been discussed and agreed upon.
- [x] The component design has been compared to the **Figma design** to ensure consistency.
- [x] The **component is fully responsive** and behaves as expected across different screen sizes.
- [x] The component **supports keyboard web accessibility** (if applicable).
- [x] I have written **unit tests** or **integration tests** to cover the component's functionality above the coverage threshold.
- [x] I have updated the **documentation** for the component, including usage examples and guidelines.
- [x] I have **commented** my code, particularly in complex or non-obvious areas.
- [x] The **component is properly exported** for use in other parts of the application.

---

## Related Issues

<!-- Link any related issue or task. For example: -->
Fixes #76

alebg commented on packages/ui-kit/lib/components/carousel/general-card-carousel.tsx:
```diff
@@ -0,0 +1,202 @@
+                            : "w-1/3"
+                        }`}
+                      >
+                        <GeneralCard
```
Hi @ObitoDevv thanks a lot, so main remark here, what's the reason you decided to do this, as opposed to pass childred as props as discussed yesterday?

alebg commented on packages/ui-kit/lib/components/carousel/generalcard.tsx:
```diff
@@ -0,0 +1,106 @@
+  const shouldShowPlaceholder = !imageUrl || isImageError;
+
+  const buttonCallback = () => {
+    window.location.href = buttonUrl;
```
@ObitoDevv About [your question in the issue](https://github.com/dream-aim-deliver/dad-e-class-monorepo/issues/76#issuecomment-2744969138), since you're using an href, your component should redirect to any url you pass on click (relative or not)

alebg commented on packages/ui-kit/stories/carousel/general-card.stories.tsx:
```diff
@@ -0,0 +1,125 @@
+    buttonText: {
+      control: 'text',
+      description: 'Text for the button'
+    },
```
@ObitoDevv if you enable the buttonUrl prop here (as I did locally), you can test and see that the button works with any url already

alebg reviewed (commented):
@ObitoDevv Please address my question about not designing the carousel as a layout component first

ObitoDevv commented:
> @ObitoDevv Please address my question about not designing the carousel as a layout component first

Hi @alebg sir, thanks for your feedback! Initially, I was exploring a more generic carousel as a layout component accepting `children`, but since the current requirement was specific to `GeneralCard` components, I shifted focus to optimize for that use case. Handling the card data directly allowed me to streamline the grouping logic and integrate translations efficiently. I created this PR to get the code reviewed and ensure it aligns with our immediate needs. That said, I see the benefit of a reusable layout component and am happy to refactor if we want to prioritize broader flexibility moving forward. Let me know your thoughts on this or if we should adjust it now!

alebg commented:
> > @ObitoDevv Please address my question about not designing the carousel as a layout component first
> 
> Hi @alebg sir, thanks for your feedback! Initially, I was exploring a more generic carousel as a layout component accepting `children`, but since the current requirement was specific to `GeneralCard` components, I shifted focus to optimize for that use case. Handling the card data directly allowed me to streamline the grouping logic and integrate translations efficiently. I created this PR to get the code reviewed and ensure it aligns with our immediate needs. That said, I see the benefit of a reusable layout component and am happy to refactor if we want to prioritize broader flexibility moving forward. Let me know your thoughts on this or if we should adjust it now!

@ObitoDevv thanks for your answer. So we might have a misunderstanding here. I mean 'layout' component in the sense explained yesterday by Mayank, related to the children as props strategy. The main argument in favor of that approach is not really about reusability, but about performance (prevent unnecessary re-renders of the components that include smaller ones inside). [This is the guide I shared yesterday about this topic](https://www.developerway.com/posts/react-re-renders-guide#part3.3), and here you can find [its video counterpart](https://youtu.be/7sgBhmLjVws?si=CgE-up_l6ZFFa4Ep&t=275).

ObitoDevv commented:
Hi @alebg, thanks for the clarification! I totally get what you mean by 'layout' component in the context Mayank explained yesterday—focusing on performance optimization rather than just reusability. In this updated version of the Carousel, I’ve implemented the "children as props" strategy to prevent unnecessary re-renders, as outlined in the guide you shared. Specifically, I split the component into CarouselContent (a memoized renderer for the items) and CarouselController (handling state and navigation). This way, the heavy lifting of rendering the children is isolated in CarouselContent, which only re-renders when its props (items or itemsPerView) change, while CarouselController manages the state updates like currentPage without triggering re-renders of the child components. I’ve kept the grouping logic and responsiveness intact, optimizing for the GeneralCard use case but still maintaining flexibility. Let me know if this aligns with the performance focus you’re aiming for, or if there’s anything else I should tweak!

alebg commented on packages/ui-kit/lib/components/carousel/carousel.tsx:
```diff
@@ -0,0 +1,231 @@
+              variant="primary"
+              size="big"
+              text={dictionary.components.generalCard.buttonText}
+              onClick={onClick || (() => alert("CTA Button Clicked"))}
```
@ObitoDevv please remove this hard coded value

carlapbg reviewed (changes requested):
Nice job! Just one thing: Can you add more space between the title and the description? It should be about the same distance as between the description and the title.

<img width="468" alt="Screenshot 2025-03-23 at 14 44 51" src="https://github.com/user-attachments/assets/ff72ad67-6ebd-4d4b-a75d-291abae7eaa7" />

carlapbg reviewed (approved):
Looks nice to me. Thank you, great job!

alebg reviewed (approved):
@ObitoDevv looks good, thanks a lot!
